File: cfg-reader/emitter.py
from .node import *
from .detail import *


class Emitter:
    def __init__(self) -> object:
        self.stream = ""

    def emit(self, node: Node, indent: int, is_last: bool = True, use_commas: bool = False, use_apostrophes: bool = False):
        _indent = ' ' * (indent * 2)

        if node.is_scalar():
            string: str = node.to_string()
            special_chars = string.count(" ")
            if use_apostrophes or special_chars > 0:
                self.stream += "'" + escape(string) + "'\n"
            else:
                self.stream += escape(string) + "\n"

        elif node.is_list():
            self.stream += "[\n"

            _list = node.to_list()
            i = 0
            end = len(_list) - 1
            for node in _list:
                self.stream += _indent
                self.emit(node, indent + 1, i == end, use_commas, use_apostrophes)
                i += 1

            self.stream += " " * ((indent - 1) * 2)

            if is_last or not use_commas:
                self.stream += "]\n"
            else:
                self.stream += "],\n"
        elif node.is_dict():
            if indent > 0:
                self.stream += "{\n"

            _dict = node.to_dict()
            i = 0
            end_idx = len(_dict) - 1

            for key, node in _dict:
                if node is None or node.is_none():
                    continue
                self.stream += _indent + key + ":"
                self.emit(node, indent + 1, i == end_idx, use_commas, use_apostrophes)
                i += 1

            if indent > 0:
                self.stream += " " * ((indent - 1) * 2)
                if is_last or not use_commas:
                    self.stream += "}\n"
                else:
                    self.stream += "},\n"
        return None

    def get_string(self):
        return self.stream

File: src/emitter.ts
import { Writable } from 'stream';
import { ConfigValue } from '.';
import { Detail } from './detail';
import { Dict, List, Node, NodeType, Scalar } from "./node";

export class Emitter {
    public stream: string = "";

    protected containsSpecials(value: string): boolean {
        return /[:,'"\[\]\{\}]/gm.test(value);
    }

    public emitNode(node: Node<Dict | List | Scalar>, os: Writable, indent: number = 0, isLast: boolean = true): void {
        const _indent = ' '.repeat(indent * 2);

        if (node.type === NodeType.Scalar) {
            os.write(`'${Detail.Escape((node as Node<Scalar>).value)}',\n`);
        } else if (node.type === NodeType.List) {
            os.write('[\n');

            const list = (node as Node<List>).value;
            for (let i = 0; i < list.length; i++) {
                const it = list[i];
                os.write(_indent);
                this.emitNode(it, os, indent + 1, i == list.length - 1);
            }

            os.write(`${' '.repeat((indent - 1) * 2)}${isLast ? ']\n' : '],\n'}`);
        } else if (node.type == NodeType.Dict) {
            if (indent > 0)
                os.write('{\n');

            const dict = (node as Node<Dict>).value;
            const keys = Object.keys(dict);
            for (let i = 0; i < keys.length; i++) {
                const key = keys[i];
                if (dict[key].type == NodeType.None)
                    continue;

                os.write(_indent + key + ':');
                this.emitNode(dict[key], os, indent + 1, i == keys.length - 1);
            }

            if (indent > 0)
                os.write(`${' '.repeat((indent - 1) * 2)}${isLast ? '}\n' : '},\n'}`);
        }
    }

    public emitConfigValue(value: ConfigValue, indent: number = 0, isLast: boolean = true, commas: boolean = true, apostrophes: boolean = true): void {
        const _indent = ' '.repeat(indent * 2);

        if (value instanceof Array) {
            //os.write('[\n');
            this.stream += '[\n';
            for (let i = 0; i < value.length; i++) {
                //os.write(_indent);
                this.stream += _indent;
                this.emitConfigValue(value[i], indent + 1, i == value.length - 1);
            }

            //os.write(_indent.repeat(indent - 1) + `${isLast || !commas ? ']\n' : '],\n'}`);
            this.stream += _indent.repeat(indent - 1) + `${isLast || !commas ? ']\n' : '],\n'}`;
        } else if (value instanceof Object) {
            if (indent > 0)
                //os.write('{\n');
                this.stream += '{\n';

            const keys = Object.keys(value);
            for (let i = 0; i < keys.length; i++) {
                const key = keys[i];
                const _value = value[key];

                if (_value == null)
                    continue;

                //os.write(_indent + key + ':');
                this.stream += _indent + key + ':';
                this.emitConfigValue(_value, indent + 1, i == keys.length - 1);
            }

            if (indent > 0)
                //os.write(_indent.repeat(indent - 1) + `${isLast || !commas ? '}\n' : '},\n'}`);
                this.stream += _indent.repeat(indent - 1) + `${isLast || !commas ? '}\n' : '},\n'}`;
        } else {
            let escaped;

            if (typeof value === "boolean") {
                escaped = Detail.Escape(String(value));
            } else if (typeof value === "string") {
                escaped = Detail.Escape(value);
            } else if (typeof value === "number") {
                escaped = Detail.Escape(value.toString());
            }

            if (escaped === undefined) {
                throw new Error(`[CFG-READER] can not emit value of type: ${typeof value}. (you passed an invalid data type)`);
            }

            const useApostrophes = apostrophes || this.containsSpecials(escaped);

            this.stream += (useApostrophes ? "'" : '') + escaped + (useApostrophes ? "'" : '') + (commas ? ',' : '') + '\n';
        } 
    }
}